minor performance improvements to url_parse (#429)

DyfanJones opened the pull request:
This is a minor performance enhancement to function `url_parse`

Issue: https://github.com/r-lib/httr2/issues/429


``` r
url <- 'https://someurl.com/with/query_string?i=main&mode=front&sid=12ab&enc=+Hello'

(bm <- bench::mark(
  httr = httr::parse_url(url),
  httr2 = httr2::url_parse(url),
  new_mthd = url_parse(url), # new method in PR
  urltools = urltools::url_parse(url),
  adaR = adaR::ada_url_parse(url),
  check = F
))
#> # A tibble: 5 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 httr          239µs    256µs     3774.  387.71KB     35.5
#> 2 httr2         207µs    232µs     4100.  466.62KB     34.8
#> 3 new_mthd      129µs    144µs     6509.     4.5MB     39.2
#> 4 urltools      126µs    134µs     7024.    7.45MB     43.8
#> 5 adaR          189µs    202µs     4755.    1.06MB     47.1

bm |> ggplot2::autoplot()
#> Loading required namespace: tidyr
```

![](https://i.imgur.com/kCjxbgG.png)<!-- -->

``` r

urls <- list(
  "/",
  "//google.com",
  "file:///",
  "http://google.com/",
  "http://google.com/path",
  "http://google.com/path?a=1&b=2",
  "http://google.com:80/path?a=1&b=2",
  "http://google.com:80/path?a=1&b=2#frag",
  "http://google.com:80/path?a=1&b=2&c=%7B1%7B2%7D3%7D#frag",
  "http://user@google.com:80/path?a=1&b=2",
  "http://user:pass@google.com:80/path?a=1&b=2",
  "svn+ssh://my.svn.server/repo/trunk",
  'https://someurl.com/with/query_string?i=main&mode=front&sid=12ab&enc=+Hello',
  "http://user:pass@example.com:80/path?a=1&b=2&c={1{2}3}#frag"
)

(bm <- bench::mark(
  httr = lapply(urls, httr::parse_url),
  httr2 = lapply(urls, httr2::url_parse),
  adaR = lapply(urls, adaR::ada_url_parse),
  urltools = lapply(urls, urltools::url_parse),
  new_mthd = lapply(urls, url_parse), # new method in PR
  check = F,
  iterations = 1000
))
#> # A tibble: 5 × 6
#>   expression      min   median `itr/sec` mem_alloc `gc/sec`
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl>
#> 1 httr          2.5ms   2.55ms      388.   161.9KB     18.3
#> 2 httr2        2.25ms   2.31ms      430.    70.9KB     20.7
#> 3 adaR         2.68ms   2.78ms      358.    34.9KB     24.5
#> 4 urltools     1.73ms    1.8ms      551.    34.9KB     23.5
#> 5 new_mthd     1.59ms   1.66ms      595.    62.9KB     22.2
bm |> ggplot2::autoplot()
```

![](https://i.imgur.com/MmKgKXT.png)<!-- -->

<sup>Created on 2024-01-25 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

hadley commented on R/parse.R:
```diff
@@ -77,7 +78,10 @@ parse_in_half <- function(x, pattern) {
-  map(pieces, empty_to_null)
+
+  # replace empty element with null
+  pieces[pieces == ""] <- list(NULL)
```
This is a bit confusing to me because it implicitly converts the character vector to a list. Maybe this would be clearer?

```R
empty <- pieces == ""
pieces <- as.list(pieces)
pieces[empty] <- list(NULL)
```

Or maybe use `ifelse()`?

```R
ifelse(pieces == "", list(NULL), as.list(pieces))
```

Would you mind seeing if those approaches work and what the impact on the benchmarks is?

DyfanJones replied:
```r
empty <- pieces == ""
pieces <- as.list(pieces)
pieces[empty] <- list(NULL)
```
Had a small (hardly noticeable) impact on performance to the implicit conversion of character vector to a list.


```r
ifelse(pieces == "", list(NULL), as.list(pieces))
```
Sadly had a larger impact and gave performance similar to the original implementation.

hadley replied:
Yeah, lets go with the explicit coercion to a list then; I think that's easier to reason about.

hadley commented on R/parse.R:
```diff
@@ -52,17 +52,18 @@ parse_delim <- function(x, delim, quote = "\"", ...) {
-  loc <- regexpr("=", x, fixed = TRUE)
-  pieces <- regmatches(x, loc, invert = TRUE)
+  if (length(x) == 0) return(NULL)
+  pieces <- strsplit(x, "=")
```
What happens if there are multiple equals in the string? It doesn't look like that case is currently covered in the tests, so you might want to add a tests specifically for `parse_name_equals_value()` too.

DyfanJones replied:
The only case it doesn't work is this type of scenario: `foo==&bar=test`. To overcome with we can do a few options:

Original Method:
```r
loc <- regexpr("=", x, fixed = TRUE)
regmatches(x, loc, invert = TRUE)
```

stringi method:
```r
stringi::stri_split_fixed(x, "=", n = 2)
```

substitution method:
```r
#i.e.
safe = "&"
strsplit(sub("=", safe, x, fixed = T), safe, fixed = TRUE)
```
**NOTE:** As `parse_name_equals_value` is called after `parse_delim` we can use the splitting delimiter in `parse_delim` as a "safe" character for the substitution method.


### Benchmark example:
``` r
safe <- "&"
query <- c(
  "foo==", "foo==&bar==", "foo==&bar==&cho==", "foo==&bar==&cho==&zoo=="
)
query_param <- lapply(query,
  httr2:::parse_delim, delim = safe
)

bms <- lapply(seq_along(query_param), \(i) {
  x <- query_param[[i]]
  bm <- bench::mark(
    original = {
      loc <- regexpr("=", x, fixed = TRUE)
      regmatches(x, loc, invert = TRUE)
    },
    stringi = stringi::stri_split_fixed(x, "=", n = 2),
    substitution = strsplit(sub("=", safe, x, fixed = T), safe, fixed = TRUE)
  )
  bm$param_no <- i
  bm
})

bms |> dplyr::bind_rows() |> ggplot2::autoplot()
#> Loading required namespace: tidyr
```

![](https://i.imgur.com/gJwyoK8.png)<!-- -->

<sup>Created on 2024-01-26 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

The substitution method has some really good performance with it only being slightly slower than `stringi`.

hadley reviewed (commented):
Thanks for working on this!

Would you mind also including before and after benchmarks of the functions specifically affected here?

DyfanJones commented:
Before and after benchmark

```r
urls <- list(
  "/",
  "//google.com",
  "file:///",
  "http://google.com/",
  "http://google.com/path",
  "http://google.com/path?a=1&b=2",
  "http://google.com:80/path?a=1&b=2",
  "http://google.com:80/path?a=1&b=2#frag",
  "http://google.com:80/path?a=1&b=2&c=%7B1%7B2%7D3%7D#frag",
  "http://user@google.com:80/path?a=1&b=2",
  "http://user:pass@google.com:80/path?a=1&b=2",
  "svn+ssh://my.svn.server/repo/trunk",
  'https://someurl.com/with/query_string?i=main&mode=front&sid=12ab&enc=+Hello',
  "http://user:pass@example.com:80/path?a=1&b=2&c={1{2}3}#frag"
)

bm <- bench::mark(
  lapply(urls, httr2::url_parse),
  iterations = 1000
)


#> # A tibble: 2 × 13
#>   expression      min   median `itr/sec` mem_alloc `gc/sec` n_itr  n_gc total_time result      memory              time               gc                  
#>   <bch:expr> <bch:tm> <bch:tm>     <dbl> <bch:byt>    <dbl> <int> <dbl>   <bch:tm> <list>      <list>              <list>             <list>              
#> 1 before       2.24ms   2.32ms      421.    62.9KB     46.8   900   100      2.14s <list [14]> <Rprofmem [27 × 3]> <bench_tm [1,000]> <tibble [1,000 × 3]>
#> 2 after        1.65ms   1.72ms      573.    62.9KB     20.8   965    35      1.68s <list [14]> <Rprofmem [28 × 3]> <bench_tm [1,000]> <tibble [1,000 × 3]>
```

![image](https://github.com/r-lib/httr2/assets/23235152/d692d7d5-732d-49be-9609-1a3e0c5894a9)

I am guessing there will be micro performance enchancements across httr2 as `parse_name_equals_value` is the main function that has been changed.

hadley commented on R/parse.R:
```diff
@@ -77,10 +83,14 @@ parse_in_half <- function(x, pattern) {
+
+  empty <- pieces == ""
+  pieces <- as.list(pieces)
+  # replace empty element with null
```
I think you can drop the comment as that's what the code expresses.

hadley commented on R/parse.R:
```diff
@@ -51,18 +51,24 @@ parse_delim <- function(x, delim, quote = "\"", ...) {
+  if (length(x) == 0) {
+    return(NULL)
+  }
+  # Note: quotes are removed in the parse_delim function
+  # re-introduce quotes to create a safe pattern to split on
```
Why do we need quotes and `&` to create the safe pattern?

I think it would also be useful to add a comment pointing out that we're replacing the first instance of `=` with a sequence that we know doesn't exist in the string to ensure that we always produce at most two pieces.

hadley commented on R/parse.R:
```diff
@@ -51,18 +51,24 @@ parse_delim <- function(x, delim, quote = "\"", ...) {
-  name <- trimws(map_chr(pieces, "[[", 1))
-
-  set_names(as.list(val), name)
+  found <- pieces_matrix[, 1] == pieces_matrix[, 2]
```
This happens because of recycling in `rbind()`?

DyfanJones replied:
Yes for example:

``` r
ll <- list(
  c("var1", "foo"),
  c("var2"),
  c("var3", "bar")
)

do.call(rbind, ll)
#>      [,1]   [,2]  
#> [1,] "var1" "foo" 
#> [2,] "var2" "var2"
#> [3,] "var3" "bar"
```

<sup>Created on 2024-01-29 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

hadley reviewed (commented):
Thanks for continuing to work on this 😄 

For the benchmarks, I meant can you please benchmark `parse_name_equals_value()` and `parse_match()` by themselves? Just so I get a better understanding of what was causing the slowness before.

DyfanJones commented:
Ah sorry about the benchmarks, my mistake:


``` r
parse_match <- function(x, pattern) {
  m <- regexec(pattern, x, perl = TRUE)
  pieces <- regmatches(x, m)[[1]][-1]
  empty <- pieces == ""
  pieces <- as.list(pieces)
  pieces[empty] <- list(NULL)
  return(pieces)
}

urls <- list(
  "/",
  "//google.com",
  "file:///",
  "http://google.com/",
  "http://google.com/path",
  "http://google.com/path?a=1&b=2",
  "http://google.com:80/path?a=1&b=2",
  "http://google.com:80/path?a=1&b=2#frag",
  "http://google.com:80/path?a=1&b=2&c=%7B1%7B2%7D3%7D#frag",
  "http://user@google.com:80/path?a=1&b=2",
  "http://user:pass@google.com:80/path?a=1&b=2",
  "svn+ssh://my.svn.server/repo/trunk",
  'https://someurl.com/with/query_string?i=main&mode=front&sid=12ab&enc=+Hello',
  "http://user:pass@example.com:80/path?a=1&b=2&c={1{2}3}#frag"
)
pattern <- "^(([^:/?#]+):)?(//([^/?#]*))?([^?#]*)(\\?([^#]*))?(#(.*))?"
bench::mark(
  before = lapply(urls, httr2:::parse_match, pattern = pattern),
  after = lapply(urls, parse_match, pattern = pattern)
) |> ggplot2::autoplot() + ggplot2::labs(title = "parse_match benchmark")
#> Loading required namespace: tidyr
```
![image](https://github.com/r-lib/httr2/assets/23235152/55e2efbc-e16f-4fc8-802f-66a154aab32d)

<sup>Created on 2024-01-29 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

`parse_match` has a micro performance improvement.

``` r
library(rlang)
parse_name_equals_value <- function(x, safe = "&", quote = "\"") {
  if (length(x) == 0) {
    return(NULL)
  }
  # Note: quotes are removed in the parse_delim function
  # re-introduce quotes to create a safe pattern to split on
  safe_pattern <- paste0(quote, safe, quote)
  pieces <- strsplit(sub("=", safe_pattern, x, fixed = T), safe_pattern, fixed = TRUE)
  pieces_matrix <- do.call(rbind, pieces)
  
  if (ncol(pieces_matrix) == 1) {
    pieces_matrix <- cbind(pieces_matrix, rep("", nrow(pieces_matrix)))
  }
  
  # If only one piece, assume it's a field name with empty value
  found <- pieces_matrix[, 1] == pieces_matrix[, 2]
  pieces_matrix[found, 2] <- ""
  set_names(as.list(pieces_matrix[, 2]), pieces_matrix[, 1])
}

params <- "foo==&var1=12ab,var2=;var3=,&var4=\"&\"var5=;out"
bench::mark(
  before = httr2:::parse_name_equals_value(httr2:::parse_delim(params, "&")),
  after = parse_name_equals_value(httr2:::parse_delim(params, "&"), "&")
) |> ggplot2::autoplot() + ggplot2::labs(title = "parse_name_equals_value benchmark")
#> Loading required namespace: tidyr
```

![image](https://github.com/r-lib/httr2/assets/23235152/5de01334-85a3-4245-9296-0d40f48228ae)

<sup>Created on 2024-01-29 with [reprex v2.1.0](https://reprex.tidyverse.org)</sup>

DyfanJones commented:
@hadley I have added in some extra comments and I have restructured `parse_name_equals_value` function. I am hoping it makes more sense in the logic it uses (note these changes have no impact on the current performance).

hadley commented on R/parse.R:
```diff
@@ -51,18 +51,35 @@ parse_delim <- function(x, delim, quote = "\"", ...) {
-  loc <- regexpr("=", x, fixed = TRUE)
-  pieces <- regmatches(x, loc, invert = TRUE)
-
+parse_name_equals_value <- function(x, safe = "&", quote = "\"") {
```
I think I've figured out a faster way to do this which is also a little clearer (if more verbose): https://gist.github.com/hadley/2fd32cdc053099d62d56f5f28898ec95

The main insight I had is that main culprit for the slow speed in my approach is `regmatches()`, and for this simple case, it's easy enough to do by hand. I also named the function `parse_in_half()` as there's an existing function with that name that only works for scalars.

DyfanJones replied:
Plus it looks like you get around the R magic from the `do.call(rbind, nested_list)`.
